src/ast/seq_decl_plugin.cpp

```diff
@@ -1595,3 +1576,108 @@ std::ostream& seq_util::re::pp::display(std::ostream& out) const {
         // Else: derivative or is_of_pred
         return out << mk_pp(e, re.m);
 }
+
+/*
+  Returns true iff info has been computed for the regex r
+*/
+bool seq_util::re::has_valid_info(expr* r) {
+    return r->get_id() < m_infos.size() && m_infos[r->get_id()].is_valid();
+}
+
+/*
+  Get the information value associated with the regular expression e
+*/
+seq_util::re::info seq_util::re::get_info(expr* e) 
+{
+    SASSERT(u.is_re(e));
+    if (!has_valid_info(e)) {
+        m_info_pinned.push_back(e);
+        compute_info_rec(e);
+    }
+    return m_infos[e->get_id()];
+}
+
+/*
+  Computes the info value for the given regex e recursively over the structure of e
+*/
+void seq_util::re::compute_info_rec(expr* e) {
+    if (has_valid_info(e))
+        return;
+
+    SASSERT(is_app(e));
+    app* ea = to_app(e);
+    SASSERT(!m.is_ite(e));
+    SASSERT(ea->get_decl()->get_decl_kind() != OP_RE_DERIVATIVE);
+    SASSERT(ea->get_decl()->get_decl_kind() != _OP_RE_ANTIMOROV_UNION);
+
+    info e_info(invalid_info);
+    unsigned k1, k2;
+    switch (ea->get_decl()->get_decl_kind())
+    {
+    case OP_RE_EMPTY_SET:
+        e_info = info(UINT_MAX);
+        break;
+    case OP_RE_FULL_SEQ_SET:
+    case OP_RE_STAR:
+    case OP_RE_OPTION:
+        e_info = info(0);
+        break;
+    case OP_RE_RANGE:
+    case OP_RE_FULL_CHAR_SET:
+    case OP_RE_OF_PRED:
+        e_info = info(1);
+        break;
+    case OP_RE_CONCAT:
+        compute_info_rec(ea->get_arg(0));
```

Comment:
if you have this return an info, then you can combine many of the lines that first compute info, then access it.